### PR TITLE
Improve unphysical check

### PR DESCRIPTION
**Pull request by hyschive**
- `Hydro_CheckUnphysical()`
  - Rename it as `Hydro_IsUnphysical()` to clarify the returned value
  - Replace `<=` with `<` and `>=` with `>` whenever appropriate to make different modes more consistent
  - `UNPHY_MODE_SING` mode: input the accepted range
  - `UNPHY_MODE_CONS` mode:
    - Check internal energy and pressure.
    - Allow internal energy to be slightly negative if it's within machine precision. This helps alleviate some false alarms in, for example, `Interpolate_Iterate()`, for extremely cold gas.
    - Pressure cannot be negative.
- `Interpolate_Iterate()`
  - Conserved variables: allow internal energy to be slightly negative if it's within machine precision (by using `Hydro_IsUnphysical()`)
  - Primitive variables: neither internal energy nor pressure can be negative
- `Typedef.h`
  - Rename `CheckUnphysical_t` as `IsUnphyMode_t`
  - Rename `VerbosityLevelUnphy_t` as `IsUnphVerb_t`
- No longer check the sound speed for `OPT__CK_INPUT_FLUID`
- For all EoS sanity checks, mass density cannot be zero but other quantities such as internal energy, pressure, temperature, and sound speed can be zero.

### CODE REVIEW

File: src/Interpolation/Interpolate.cpp [hunk 1]
```diff
@@ -404,13 +404,22 @@ void Interpolate_Iterate( real CData[], const int CSize[3], const int CStart[3],
...
+                  if ( Eint != NULL_REAL )   Aux_Message( stderr, "Eint=%14.7e\n", Eint );
+               }
+
+               else {
+                  const real CheckMinPres_No = false;
+                  const real Pres = Hydro_Con2Pres( Temp[DENS], Temp[MOMX], Temp[MOMY], Temp[MOMZ], Temp[ENGY], Temp+NCOMP_FLUID,
+                                                    CheckMinPres_No, NULL_REAL, Emag,
+                                                    EoS_DensEint2Pres_CPUPtr, EoS_GuessHTilde_CPUPtr, EoS_HTilde2Temp_CPUPtr,
+                                                    EoS_AuxArray_Flt, EoS_AuxArray_Int, h_EoS_Table, &Eint );
+                  Aux_Message( stderr, "Eint=%14.7e, Pres=%14.7e\n", Eint, Pres );
```
Comment (hfhsieh) [inline src/Interpolation/Interpolate.cpp:422]: In the case `FData_is_Prim = false`, `Eint` is not recalculated yet.

**Review by hfhsieh — APPROVED** [review]
Looks good to me. I've added some comments for minor issues.

File: src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp [hunk 2]
```diff
@@ -795,34 +803,48 @@ real Hydro_CheckMinEintInEngy( const real Dens, const real MomX, const real MomY
...
-//                   UNPHY_MODE_SING         : Check if the input single field is NAN, inf or negative
+//                   UNPHY_MODE_SING         : Check if the input single field is NAN or lies outside the accepted range
 //                   UNPHY_MODE_CONS         : Check if the input conserved variables, including passive scalars, are unphysical
 //                   UNPHY_MODE_PRIM         : Check if the input primitive variables, including passive scalars, are unphysical
 //                   UNPHY_MODE_PASSIVE_ONLY : Check if the input passive scalars are unphysical
 //                2. For UNPHY_MODE_CONS with SRHD, we also check if Eq. 15 in "Tseng et al. 2021, MNRAS, 504, 3298"
 //                   has a positive root
-//                3. UNPHY_MODE_CONS currently does not check gas pressure
+//                3. For UNPHY_MODE_CONS:
+//                   - Mass density and total energy density must be postive
```
[the remark this answers is not in the document]
Comment (hyschive) [reply src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp:817]: Fixed

File: src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp [hunk 3]
```diff
@@ -832,168 +854,221 @@ bool Hydro_CheckUnphysical( const CheckUnphysical_t Mode, const real Fields[], c
...
 //       print out the unphysical values
 #        if ( !defined __CUDACC__  ||  defined CHECK_UNPHYSICAL_IN_FLUID )
-         if ( FailCell && Verbose )
+         if ( UnphyCell && Verbose )
          {
-            printf( "ERROR: unphysical primitive variables at file <%s>, line <%d>, function <%s> !!\n",
+            printf( "ERROR : unphysical primitive variables at file <%s>, line <%d>, function <%s> !!\n",
                     File, Line, Function );
-            printf( "D=%14.7e, Vx=%14.7e, Vy=%14.7e, Vz=%14.7e, P=%14.7e\n",
+            printf( "D=%14.7e, Vx=%14.7e, Vy=%14.7e, Vz=%14.7e, P=%14.7e",
```
[the remark this answers is not in the document]
Comment (hyschive) [reply src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp:1036]: Fixed

File: src/EoS/User_Template/CPU_EoS_User_Template.cpp [hunk 4]
```diff
@@ -241,7 +257,9 @@ static real EoS_DensPres2CSqr_User_Template( const real Dens, const real Pres, c
 
 // check
 #  ifdef GAMER_DEBUG
-   if ( Hydro_CheckUnphysical( UNPHY_MODE_SING, &Cs2, "output sound speed squared", ERROR_INFO, UNPHY_VERBOSE ) )
+   if (  Hydro_IsUnphysical( UNPHY_MODE_SING, &Cs2, "output sound speed squared",
+                             (real)0.0, HUGE_NUMBER, NULL_REAL, NULL, NULL, NULL, NULL, NULL, NULL,
```
[the remark this answers is not in the document]
Comment (hyschive) [reply src/EoS/User_Template/CPU_EoS_User_Template.cpp:261]: Sound speed can be zero in general due to the limited floating-point accuracy. But it should be positive for the isothermal EoS, where it is directly proportional to the input temperature `ISO_TEMP`.

File: src/Interpolation/Interpolate.cpp [hunk 5]
```diff
@@ -355,31 +367,19 @@ void Interpolate_Iterate( real CData[], const int CSize[3], const int CStart[3],
...
-#              endif
-
-               Pres = Hydro_Con2Pres( Temp[DENS], Temp[MOMX], Temp[MOMY], Temp[MOMZ], Temp[ENGY], Temp+NCOMP_FLUID,
-                                      CheckMinPres_No, NULL_REAL, Emag,
-                                      EoS_DensEint2Pres_CPUPtr, EoS_GuessHTilde_CPUPtr, EoS_HTilde2Temp_CPUPtr,
-                                      EoS_AuxArray_Flt, EoS_AuxArray_Int, h_EoS_Table, NULL );
-
-               if (  Hydro_CheckUnphysical( UNPHY_MODE_SING, &Pres, "interpolated pressure", ERROR_INFO, UNPHY_SILENCE )  )
-                  Fail_ThisCell = true;
+//             one can add addtional checks for conserved variables here
```
[the remark this answers is not in the document]
Comment (hyschive) [reply src/Interpolation/Interpolate.cpp:382]: Fixed

File: src/Interpolation/Interpolate.cpp (hunk 1 above)
Comment (hyschive) [reply src/Interpolation/Interpolate.cpp:422]: It is also returned by `Hydro_Con2Pres()` from the last argument.

**Comment by hyschive** [comment]
@hfhsieh Thanks for the careful review! I've addressed all your comments and will merge this PR right away.

**Pull request merged**